pkg/reaper/nodereaper/helpers.go

```diff
@@ -209,6 +227,124 @@ func (ctx *ReaperContext) publishEvent(namespace string, event *v1.Event) error
 	return nil
 }
 
+func (ctx *ReaperContext) obtainReapLock(ddbAPI dynamodbiface.DynamoDBAPI, nodeName, instanceID, nodeType string) (LockRecord, error) {
+	log.Infof("obtaining lock for a %s node %s (%s)", nodeType, nodeName, instanceID)
+
+	timestamp := time.Now().Format(time.RFC3339)
+
+	lock := LockRecord{
+		LockType:   nodeType,
+		ClusterID:  ctx.ClusterID,
+		NodeName:   nodeName,
+		InstanceID: instanceID,
+		CreatedAt:  timestamp,
+		tableName:  ctx.LocksTableName,
+	}
+
+	err := lock.obtainLock(ddbAPI)
+	return lock, err
+}
+
+func (l LockRecord) obtainLock(ddbAPI dynamodbiface.DynamoDBAPI) error {
+	serializedLock, err := dynamodbattribute.MarshalMap(l)
+	if err != nil {
+		return err
+	}
+	input := &dynamodb.PutItemInput{
+		Item:                serializedLock,
+		TableName:           aws.String(l.tableName),
+		ConditionExpression: aws.String("attribute_not_exists(LockType)"),
+	}
+
+	_, err = ddbAPI.PutItem(input)
+	if err != nil {
+		log.Infof("failed to obtain lock for a %s node %s (%s): %s", controlPlaneType, l.NodeName, l.InstanceID, err.Error())
+		return err
+	}
+
+	l.locked = true
+
+	log.Infof("successfully obtained lock for a %s node %s (%s)", controlPlaneType, l.NodeName, l.InstanceID)
+
+	return err
+}
+
+// TODO: should this be (ddbAPI, lock) instead? Or Lock.tryClearLock(ddbAPI)?
```

Comment:
Considering that you already split construction of lock and performing the actual locking, I like `Lock.tryClearLock(ddbAPI)`

Reply:
Oh you actually didn't, `obtainReapLock` calls `obtainLock`. In that case, the alternatives rely on having the necessary params set even if locking failed which is fine, I think.

Reply:
Lock object _should_ have all the right values even when locking fails, yes.